shell/platform/fuchsia/flutter/gfx_session_connection.cc

```diff
@@ -227,43 +228,53 @@ GfxSessionConnection::GfxSessionConnection(
   // Set the |fuchsia::ui::scenic::OnFramePresented()| event handler that will
   // fire every time a set of one or more frames is presented.
   session_wrapper_.set_on_frame_presented_handler(
-      [this](fuchsia::scenic::scheduling::FramePresentedInfo info) {
-        std::lock_guard<std::mutex> lock(mutex_);
+      [thiz = this,  // Forces us to be explicit with |this|. Prefer |weak|.
+       weak = weak_factory_.GetWeakPtr()](
+          fuchsia::scenic::scheduling::FramePresentedInfo info) {
+        if (!weak) {
+          return;
+        }
+
+        std::lock_guard<std::mutex> lock(weak->mutex_);
 
         // Update Scenic's limit for our remaining frames in flight allowed.
         size_t num_presents_handled = info.presentation_infos.size();
 
         // A frame was presented: Update our |frames_in_flight| to match the
         // updated unfinalized present requests.
-        frames_in_flight_ -= num_presents_handled;
+        weak->frames_in_flight_ -= num_presents_handled;
 
         TRACE_DURATION("gfx", "OnFramePresented5", "frames_in_flight",
-                       frames_in_flight_, "max_frames_in_flight",
-                       kMaxFramesInFlight, "num_presents_handled",
+                       weak->frames_in_flight_, "max_frames_in_flight",
+                       weak->kMaxFramesInFlight, "num_presents_handled",
                        num_presents_handled);
-        FML_DCHECK(frames_in_flight_ >= 0);
+        FML_DCHECK(weak->frames_in_flight_ >= 0);
 
-        last_presentation_time_ = fml::TimePoint::FromEpochDelta(
+        weak->last_presentation_time_ = fml::TimePoint::FromEpochDelta(
             fml::TimeDelta::FromNanoseconds(info.actual_presentation_time));
 
         // Scenic retired a given number of frames, so mark them as completed.
-        // Inspect updates must run on the inspect dispatcher.
-        async::PostTask(
-            inspect_dispatcher_, [this, now = Now(), num_presents_handled]() {
-              presents_completed_.Add(num_presents_handled);
-              last_frame_completed_.Set(now.ToEpochDelta().ToNanoseconds());
-            });
-
-        if (fire_callback_request_pending_) {
-          FireCallbackMaybe();
+        //
+        // Inspect updates must run on the inspect dispatcher. We don't pass
+        // |weak| to the callback because there is no guarantee that the
```

Comment:
I'm not sure if this statement that "there is no guarantee that callbacks [invoked by inspect_dispatcher_] will run on the same thread" is true or not. However, because I'm not certain, it feels safer to avoid passing the WeakPtr into here.

Reply:
Upon closer inspection of the code, that statement is not true n this case.  But it's all super implicit (sorry about that) so it requires scanning the entire class to know that....

1. inspect_dispatcher_ == async_get_default_dispatcher() from the thread the GfxConnection ctor ran on (this is in the constructor).
2. session_wrapper's dispatcher also == async_get_default_dispatcher() from the thread the GfxConnection ctor ran on (the fidl InterfaceHandle was bound to this thread in the constructor).
3.  Thus, the body of the session_wrapper_.set_on_frame_presented_handler callback runs on the same dispatcher as inspect_dispatcher_.  All of this code is running on the same thread.

Again, super non-obvious without a deep understanding of this class.  Suggestions on how to make that more clear very welcome.

Reply:
Further explanation:  the only part of this class that runs on another thread is the "AwaitVsync" method.   The flutter engine calls back into that method on the "UI" thread.  All of the other operations in this class are taking place on the "Raster" thread.

Reply:
No worries and thanks for the detailed explanation - super useful to know what threads different parts of this code are running on. I had a hard time finding the header for async_get_default_dispatcher, but now that I've found it, it's super clear. Changed the code to avoid the nasty `thiz = this` nonsense.